### ISSUE TITLE
On Workspace changed, new instance of AstBuilder and Profiling reseted

### ISSUE THREAD

**Issue by alvpickmans** [DynamoDS/Dynamo#9955]
## Dynamo version

2.3.0.5885

## Operating system

Windows 10

## What did you do? 
Simple example on enabling profiling on Workspace changed.

```csharp
private void OnWorkspaceChanged(IWorkspaceModel workspace)
{
    HomeWorkspaceModel homeWorkspace = workspace as HomeWorkspaceModel;

    if (homeWorkspace == null)
        return;

    homeWorkspace.EngineController.EnableProfiling(true, homeWorkspace, homeWorkspace.Nodes);
}
```

## What did you expect to see?

Profiling enabled for the newly opened workspace.

## What did you see instead?

Profiling can be enabled only when the current `EngineController`'s private `AstBuilder` property doesn't have an active `ProfilingSession`. This is fine but for the fact that when opening a new file (changing the workspace), the `EngineController` gets reseted with a new instance and subsequently, its `AstBuilder`.

Problem is that this reset happens **after** the `WorkspaceChanged` event gets fired, so by the time we can enable profiling on the subscribed event, the `EngineController` gets reseted and profiling back to be disabled.

https://github.com/DynamoDS/Dynamo/blob/master/src/DynamoCore/Models/DynamoModel.cs#L1541-L1570

```csharp
private bool OpenJsonFileFromPath(string fileContents, string filePath, bool forceManualExecutionMode)
{
    try
    {
        DynamoPreferencesData dynamoPreferences = DynamoPreferencesDataFromJson(fileContents);
        if (dynamoPreferences != null)
        {
            // TODO, QNTM-1101: Figure out JSON migration strategy
            if (true) //MigrationManager.ProcessWorkspace(dynamoPreferences.Version, xmlDoc, IsTestMode, NodeFactory))
            {
                WorkspaceModel ws;
                if (OpenJsonFile(filePath, fileContents, dynamoPreferences, forceManualExecutionMode, out ws))
                {
                    OpenWorkspace(ws); // <-- THIS FIRES 'WorkspaceChanged' EVENT

                    //Raise an event to deserialize the view parameters before
                    //setting the graph to run
                    OnComputeModelDeserialized();

                    SetPeriodicEvaluation(ws); // <-- THIS RESETS 'EngineController' AND SUBSEQUENTLY DISABLES PROFILING
                }
            }
        }
        return true;
    }
    catch (Exception e)
    {
        Console.WriteLine(e.Message);
        return false;
    }
}
```

**Comment by alvpickmans** [comment]
I guess my doubts are if this is the desired behaviour, which other means developers have to facilitate control over profiling. Happy to discuss.

**Comment by mjkkirschner** [comment]
@scottmitchell have you seen this as well? Have you worked around it somehow? 

- I guess one way to workaround it is to set the profilingEnabled on the evaluationStarted event instead of the workspace changing or any other event which happens only when a workspace is actually ready.

**Comment by alvpickmans** [comment]
Ok so playing a bit with events I managed to use `EvaluationStarted` to enable profiling with one caveat: when opening a file, nodes get compiled (`AstBuilder.CompileToAst` I think) before `OnEvaluationStarted` gets called on `HomeWorkspaceModel.Run()` method (around line 660?), meaning file needs to be run twice for profiling to work.

I also found out that falls to the developer to track if profiling is enabled and in which EngineController to avoid infinite cyclic runs in automatic mode, as `EnableProfiling` requests a new run every time.

**Comment by scottmitchell** [comment]
I'm facing the same issue currently, and do not have a workaround. I will have more comments as I get further into it.

I also agree with @alvpickmans that it falls to the developer to track if profiling is enabled, which is a bit strange.

**Comment by mjkkirschner** [comment]
@alvpickmans @scottmitchell can this be closed?

**Comment by scottmitchell** [comment]
@mjkkirschner let's leave it open for now. I'm looking at this as I set up testing for TuneUp. I imagine this will lead to a Dynamo PR.

**Comment by alvpickmans** [comment]
Agreed, still needs to be addressed. @scottmitchell it might worth to have a chat to discusse best options? I have an idea of how it should work, but you're probably more familiar with what can be changed in Dynamo
